rspace/src/main/scala/coop/rchain/rspace/HotStore.scala

```diff
@@ -29,264 +26,363 @@ trait HotStore[F[_], C, P, A, K] {
-                for {
-                  historyContinuations <- historyReader.getContinuations(channels)
-                  _ <- cacheRef.update { cache =>
-                        ignore(cache.continuations.putIfAbsent(channels, historyContinuations))
-                        cache
-                      }
-                } yield historyContinuations
-              case Some(continuations) => continuations.pure[F]
-            }
-    } yield (res)
-
   def putContinuation(channels: Seq[C], wc: WaitingContinuation[P, K]): F[Unit] =
     for {
-      continuations <- internalGetContinuations(channels)
-      _ <- cacheRef.update { cache =>
-            ignore(cache.continuations.put(channels, wc +: continuations))
-            cache
+      fromHistoryStore <- getContFromHistoryStore(channels)
+      _ <- hotStoreState.update { state =>
+            val curVal = state.continuations.getOrElse(channels, fromHistoryStore)
+            val newVal = wc +: curVal
+            state.copy(continuations = state.continuations.updated(channels, newVal))
           }
     } yield ()
 
   def installContinuation(channels: Seq[C], wc: WaitingContinuation[P, K]): F[Unit] =
-    cacheRef.update { cache =>
-      ignore(cache.installedContinuations.put(channels, wc))
-      cache
+    hotStoreState.update { cache =>
+      cache.copy(installedContinuations = cache.installedContinuations.updated(channels, wc))
     }
 
   def removeContinuation(channels: Seq[C], index: Int): F[Unit] =
     for {
-      continuations <- getContinuations(channels)
-      cache         <- cacheRef.get
-      installed     = cache.installedContinuations.get(channels)
-      _ <- if (installed.isDefined && index == 0) {
-            new IllegalArgumentException("Attempted to remove an installed continuation").raiseError
-          } else
-            checkIndex(continuations, index) >> cacheRef.update { cache =>
-              val updated = removeIndex(
-                continuations,
-                index
+      fromHistoryStore <- getContFromHistoryStore(channels)
+      r <- hotStoreState.modify[(Boolean, Boolean)](state => {
+            val curVal       = state.continuations.getOrElse(channels, fromHistoryStore)
+            val installedCon = state.installedContinuations.get(channels)
+            val isInstalled  = installedCon.nonEmpty
+
+            val removingInstalled = (isInstalled && index == 0)
+            // why -1 here
+            // from users who use `HotStore` trait perspective
+            // the users need to call getContinuations first to retrieve the continuation of the channels
+            // then find which index of the continuation users want to remove
+            // while getContinuations would return results like `Vector(installedContinuation, normalcontinuations, ...)`
+            // The first one would be always `installedContinuation` if it exist which need the
+            // index to -1 to exclude installedContinuation
```

Comment:
But index supplied here already accounts the presence of installed continuations. When you match produce over Seq of continuations, you will see the installed one, as you aways get it through `getContinuations`. I don't think you can adjust index here.

Reply:
Previously continuations are retrieved with `getContinuations` which included installed, but now they are accessed directly from the state without installed.

I'm OK with this solution for now (it's works the same as previous) until we came up with more elegant one.

Reply:
Yes I agree https://github.com/rchain/rchain/pull/3526#discussion_r761162294